src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs

```diff
@@ -282,6 +279,89 @@ public Task<ResultSetSubset> GetSubset(int startRow, int rowCount)
+            }
+
+            // Make sure there isn't a task for this file already
+            Task existingTask;
+            if (SaveTasks.TryGetValue(saveParams.FilePath, out existingTask))
+            {
+                if (existingTask.IsCompleted)
+                {
+                    // The task has completed, so let's attempt to remove it
+                    if (!SaveTasks.TryRemove(saveParams.FilePath, out existingTask))
+                    {
+                        throw new InvalidOperationException(SR.QueryServiceSaveAsMscStartingError);
+                    }
+                }
+                else
+                {
+                    // The task hasn't completed, so we shouldn't continue
+                    throw new InvalidOperationException(SR.QueryServiceSaveAsInProgress);
+                }
+            }
+
+            // Create the new task
+            Task saveAsTask = new Task(async () =>
+            {
+                try
+                {
+                    // Set row counts depending on whether save request is for entire set or a subset
+                    long rowEndIndex = RowCount;
+                    int rowStartIndex = 0;
+                    if (saveParams.IsSaveSelection)
+                    {
+                        // ReSharper disable PossibleInvalidOperationException  IsSaveSelection verifies these values exist
+                        rowEndIndex = saveParams.RowEndIndex.Value + 1;
+                        rowStartIndex = saveParams.RowStartIndex.Value;
+                        // ReSharper restore PossibleInvalidOperationException
+                    }
+
+                    using (var fileReader = fileFactory.GetReader(outputFileName))
+                    using (var fileWriter = fileFactory.GetWriter(saveParams.FilePath))
+                    {
+                        // Iterate over the rows that are in the selected row set
+                        for (long i = rowStartIndex; i < rowEndIndex; ++i)
+                        {
+                            var row = fileReader.ReadRow(fileOffsets[i], Columns);
+                            fileWriter.WriteRow(row, Columns);
+                        }
+                        if (successHandler != null)
+                        {
+                            await successHandler(saveParams);
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    fileFactory.DisposeFile(saveParams.FilePath);
+                    if (failureHandler != null)
+                    {
+                        await failureHandler(saveParams, e.Message);
+                    }
+                }
+            });
+
+            // If saving the task fails, return a failure
+            if (!SaveTasks.TryAdd(saveParams.FilePath, saveAsTask))
+            {
+                throw new InvalidOperationException(SR.QueryServiceSaveAsMscStartingError);
```

Comment:
>Msc [](start = 73, length = 3)

It's not clear to me what 'Msc' means in this context.  Maybe better to use the whole word(s) this is abbreviating